### ISSUE TITLE
Using snippets inside a repeater variable of a static page

### ISSUE THREAD

**Issue by chocolata** [rainlab/pages-plugin#336]
Hi,

I'm using the Static Pages Plugin to create different segments on a page, like so:

```
{repeater name="fields" tab="Segments" prompt="Toevoegen"}
{variable name="title" label="Title" type="text"}{/variable}
{variable name="text" label="Text" type="richeditor" mode="wysiwyg"}{/variable}
{variable name="image" label="Image" type="mediafinder" mode="image"}{/variable}
{/repeater}
```
The idea behind it, is to display the segments below eachother in a specifically set styling, as follows:

```
{% for field in fields %}
<section>
    <h2>{{ field.title }}</h2>
    {{ field.text | raw }}
    <img src="{{ field.image | media }}" alt="{{ field.title }}" />
</section>
{% endfor %}
```
This works perfectly. But I'd like to allow the content manager to use snippets in the richeditor text field. This seems like it works, but upon clicking on a snippet on the left, the snippet gets added to each instance of the repeater and on the frontend nothing displays.  In my case, snippets can be a button, or a set of contactdetails (that get their values from the theme settings).

If we could get this working, this would add even more flexibility. My workaround now is to not use repeaters, but to use set variables. As you can imagine this is a lot less clear and elegant, as reordering is not possible and there would be a fixed number of segments...

Any thoughts? Thanks in advance.

**Comment by briddle** [comment]
I can confirm this. When adding snippets to a richeditor that is part of a repeater field the snippet gets added to each instance of the repeater and on the frontend nothing displays. 

The expected behaviour is to either not allow snippets here or to make them work :(

**Comment by bennothommo** [comment]
@maartenmachiels @briddle As you need to click the Snippet in the list in order to add it to a richeditor, it will be very difficult to determine which richeditor is the intended target - especially in the case of a repeater where there could be multiple richeditors visible. A potential fix would be to allow the user to drag the snippet from the list into one of the richeditors to add it.

If you would like to see this, please feel free to submit a PR.

**Comment by maartenmachiels** [comment]
Hi Ben, thanks for your comments. Upon further inspection, I notice that snippets don't work in richeditor variables (repeater or not), but they do seem to work in richeditor placeholders.

**Comment by bennothommo** [comment]
@maartenmachiels Thanks for letting me know. I'll hopefully tackle this issue at some point.

**Comment by buffycpp** [comment]
@bennothommo here is some not-very-pretty code that i wrote sometime after i opened this issue. it fixes some stuff that i personally believe should be built-in features of staticpages. it's just a part of a plugin but you easily get the idea and maybe you or someone else will find it useful.

it will:

* allow syntax fields in content files
* allow syntax fields in partials (but not partials nested in other partials!)
* make snippets work everywhere (even in custom tabs)

again this code is super dirty because i kinda had to reverse engineer how everything works first and was just happy that i got it to work lol

`    protected function flattenTwigNode($node) {
        $result = [];
        if (!$node instanceof \Twig_Node) {
            return $result;
        }

        foreach ($node as $subNode) {
            $flatNodes = $this->flattenTwigNode($subNode);
            $result = array_merge($result, [$subNode], $flatNodes);
        }

        return $result;
    }

    public function addSyntaxFields($form, $translatableObject, $syntax) {

        $fieldsAdded = 0;
        $editorFields = [];

        try {
            $editorFields = $syntax->toEditor();
        }
        catch (\Throwable $e) {
            unset($e);            
            return $fieldsAdded;
        }

        foreach ($editorFields as $key => $fields) {
            $isSecondary = true;

            if ( array_key_exists('placement', $fields) )
            {
                if($fields['placement'] == 'primary') {
                    $isSecondary = false;
                }
            }

            //Translation support
            $translatableTypes = ['text', 'textarea', 'richeditor', 'repeater'];
            if (in_array($fields['type'], $translatableTypes)) {
                array_push($translatableObject->translatable, "viewBag[${key}]");
            }

            $finalField = null;

            if ( $fields['type'] == 'repeater' ) {
                $data = $fields;
                unset($data['fields']);
                $data['form'] = ['fields' => []];

                foreach ($fields['fields'] as $subkey => $value) {
                    $data['form']['fields']["${subkey}"] = $value;
                }

                $finalField = $data;
            }
            else 
            {
                $finalField = $fields;
            }

            //content is always in the main tab to prevent visual glitches
            if ($form->model instanceof \RainLab\Pages\Classes\Content) {
                $finalField['tab'] = 'cms::lang.editor.content';
            }

            if ( $isSecondary ) {
                $cssClass = &$finalField['cssClass'];
                if ( $cssClass != null ) {
                    $cssClass .= ' secondary-tab';
                } else {
                    $cssClass = 'secondary-tab';
                }

                //$form->addSecondaryTabFields($finalField);
                $form->secondaryTabs["fields"]["viewBag[${key}]"] = $finalField;
            } else {
                //$form->addTabFields($finalField);
                $form->tabs["fields"]["viewBag[${key}]"] = $finalField;
            }

            $fieldsAdded++;
        }

        return $fieldsAdded;
    }

    public function parseSyntaxFieldMarkup($translatableObject, $fieldValues, $syntax) {
        $viewBag = $translatableObject->getViewBag();
        if ( $viewBag != null ) {

            $translate = class_exists($translatableObject->translatableModel);

            foreach ($viewBag->getProperties() as $key => $value) {
                $fieldValues[$key] = $translate ? 
                    $translatableObject->getAttributeTranslated("viewBag[${key}]") : $value;
            }

            return [$syntax->render($fieldValues), $fieldValues];
        }
    }

    private function getTwigEnvironment($loader, $obj) {

        $twig = new Twig_Environment($loader);
        $twig->addExtension(new CmsTwigExtension());
        $twig->addExtension(new SystemTwigExtension);

        //dump($obj->getTwig());

        $isDebugMode = Config::get('app.debug', false);
        if ($isDebugMode) {            
            $twig->addExtension(new DebugExtension($obj));
        }

        return $twig;
    }

    public function boot()
    {        
        // Make partials parse syntax fields
        if ( Settings::get('parse_partials', true) ) {
            Event::listen('cms.page.renderPartial', function($obj, $name, $content) {
                $syntax = SyntaxParser::parse($content);        

                try {
                    $fieldValues = $syntax->getFieldValues();
                    if ( sizeof($fieldValues) >= 1 ) {
    
                        //has fields
                        $router = new Router(Theme::getActiveTheme());
                        $staticPage = $router->findByUrl($obj->getRouter()->getUrl());
    
                        if ( $staticPage != null ) {
                            $parsedTwig = $this->parseSyntaxFieldMarkup($staticPage, $fieldValues, $syntax);

                            $loader = new Twig_Loader_Array(array(
                                'content' => $parsedTwig[0],
                            ));

                            $twig = $this->getTwigEnvironment($loader, $obj);                            

                            return $twig->render('content', $parsedTwig[1]);
                        }
                    }
                } catch (\Throwable $e) {
                    unset($e);
                }
            });
        }

        // Make content parse syntax fields
        if ( Settings::get('parse_content', true) ) {
            Event::listen('cms.page.renderContent', function($obj, $name, $content) {        
                
                $contentFile = Content::loadCached($obj->getTheme(), $name);
                if ( $contentFile == null )
                    return;                          

                $syntax = SyntaxParser::parse($contentFile->getTwigContent());                                

                try {
                    $fieldValues = $syntax->getFieldValues();
                    if ( sizeof($fieldValues) >= 1 ) {
    
                        //has fields
                        $parsedTwig = $this->parseSyntaxFieldMarkup($contentFile, $fieldValues, $syntax);

                        $loader = new Twig_Loader_Array(array(
                            'content' => $parsedTwig[0],
                        ));
                                            
                        $twig = $this->getTwigEnvironment($loader, $obj);
                        
                        return $twig->render('content', $parsedTwig[1]);
                    }
                }
                catch(\Throwable $e) {
                    unset($e);
                }
            });
        }

        // Extend all backend form usage
        Event::listen('backend.form.extendFieldsBefore', function($form) {

            if  ( !Settings::get('global_toggle', true) )
                return;

            // Only for the User controller
            if (!$form->getController() instanceof \RainLab\Pages\Controllers\Index) {
                return;
            }

            if ( $form->isNested )
                return;

            if ( Settings::get('parse_content', true) ) {
                if ($form->model instanceof \RainLab\Pages\Classes\Content) {
                    $content = $form->model;
    
                    $syntax = SyntaxParser::parse($content->getParsedMarkupAttribute());
                    if($this->addSyntaxFields($form, $content, $syntax) > 0) {
                        //hide default tab
                        //removing causes markup to be lost
                        $form->secondaryTabs['fields']['markup_html']['cssClass'] = 'hidden';
                        $form->secondaryTabs['fields']['markup']['cssClass'] = 'hidden';
    
                        //prevent FROALA from wrapping everything in <p> tags ¯\_(ツ)_/¯
                        $form->secondaryTabs['fields']['markup_html']['type'] = 'codeeditor';
                        $form->secondaryTabs['fields']['markup']['type'] = 'codeeditor';
                    }
                }
            }

            
            if ( Settings::get('parse_partials', true) ) {
                if ($form->model instanceof \RainLab\Pages\Classes\Page) {
    
                    // Add fields for partial syntax fields
                    $staticPage = $form->model;
                    $viewBag = $staticPage->getViewBag();
    
                    $properties = $viewBag->getProperties();
                    if ( array_key_exists('layout', $properties) ) {
                        $layoutName = $viewBag->getProperties()['layout'];
                        $theme = Theme::getActiveTheme();
                        $layout = Layout::load($theme, $layoutName);
                        if ( $layout != null ) {
                            $bodyNode = $layout->getTwigNodeTree()->getNode('body')->getNode(0);
                            $nodes = $this->flattenTwigNode($bodyNode);
                    
                            foreach ($nodes as $node) {
                                if (!$node instanceof \Cms\Twig\PartialNode) {
                                    continue;
                                }
                
                                $partialName = $this->flattenTwigNode($node)[1]->getAttribute('value');
                                $partial = Partial::loadCached($theme, $partialName);
                                $syntax = SyntaxParser::parse($partial->getTwigContent());
                
                                $this->addSyntaxFields($form, $staticPage, $syntax);
                            }
            
                            $this->addDefaultFields($form, $staticPage);
                        }
                    }
                }                   
            }
        });

        if ( Settings::get('parse_snippets', true) ) {
            $reflectionBypass = new class extends \RainLab\Pages\Classes\Snippet {
                public static function extractSnippetsFromMarkup($markup, $theme) {
                    return parent::extractSnippetsFromMarkup($markup, $theme);
                }
            };
    
            Event::listen('cms.page.initComponents', function($obj, $page) use ($reflectionBypass) {
    
                $activeTheme = $obj->getTheme();
                $router = new Router($activeTheme);
                $staticPage = $router->findByUrl($obj->getRouter()->getUrl());
    
                if ( $staticPage == null )
                    return;
    
                $componentManager = ComponentManager::instance();
                if ( $componentManager == null )
                    return;
    
                foreach ($staticPage->getViewBag()->getProperties() as $key => $value) {
                    if ( is_array($value) )
                        continue;
    
                    $snippets = $reflectionBypass::extractSnippetsFromMarkup($value, $activeTheme);
                    if ( sizeof($snippets) <= 0 )
                        continue;
    
                    foreach ($snippets as $key => $data) {
    
                        if ( $data['component'] == null )
                            continue;
                    
                        if (!$componentManager->hasComponent($data['component'])) {
                            $componentManager->registerComponent($data['component'], $data['code']);
                        }                    
                        
                        $componentObj = $obj->addComponent(
                            $data['component'],
                            $data['code'],
                            $data['properties']
                        );                                        
                    }
                }
            });
    
            //Process AFTER StaticPages has processed everything and is ready to render
            Event::listen('cms.page.display', function($obj, $url, $page, $result) use($reflectionBypass) {
                $activeTheme = $obj->getTheme();
                $snippets = $reflectionBypass::extractSnippetsFromMarkup($result, $activeTheme);             
    
                foreach ($snippets as $key => $data) {
                    $replacement = '';
    
                    if ( $data['component'] == null ) {
                        $partialSnippetMap = SnippetManager::instance()->getPartialSnippetMap($activeTheme);
                        $replacement = $obj->renderPartial($partialSnippetMap[$data['code']], $data['properties']);
                    } else {
                        $replacement = $obj->renderComponent($data['code']);
                    }
    
                    $result = str_replace($key, $replacement, $result);
    
                }
    
                return $result;            
            });
        }
    }`

**Comment by bennothommo** [comment]
@buffycpp Thanks heaps for that. I'll just focus on the snippets portion for this PR, but I will definitely consider the other changes with allowing syntax fields from content and partial files going forward.

I also agree that a Froala button would definitely be useful (that's what this plugin does already: https://octobercms.com/plugin/inetis-richeditorsnippets). I still believe that having a drag-and-drop action for the snippets makes sense as the CMS pages establish that a person can add a component to a page, layout or partial by dragging it from the side panel onto the code. It even adds a `{% component %}` block in the code when you do it this way.

**Comment by buffycpp** [comment]
@bennothommo yes that plugin displays the snippet button in the richeditor but the snippets portion of it is broken because it doesn't inject the component that the snippet relies on (for non partial snippets). thus the dependency is missing and everything explodes 🔥 

another interesting thing would be to be able to decide which property should be used for snippet display in the richeditor view. consider this: https://i.imgur.com/k3JBzWU.png

each SingleView snippet has a "gallery" property that is different from the others (like seen here https://i.imgur.com/FqzQF9Y.png). it would make a lot more sense to have it display that instead of just SingleView. i guess thats a different story though 🤷‍♂️

**Comment by daftspunk** [comment]
This is fixed in October CMS v3.4 with the introduction of global snippets: https://github.com/octobercms/october/releases/tag/v3.4.0
